### ISSUE TITLE
Super X -inteligencia artificial ,(gpt/IA )

### ISSUE THREAD

**Issue by 1042422227** [google-gemini/gemini-cli#6110]
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Embedding, LayerNormalization, Dropout, Dense
from tensorflow.keras import layers

# --- 1. Preparar y tokenizar los datos ---
# Los Transformers trabajan con datos de texto convertidos a números (tokens).
# Creamos un pequeño y simple corpus de ejemplo.
corpus = [
    'El gato juega en la alfombra.',
    'El perro duerme en el sofa.',
    'La gata caza un raton.',
    'El raton corre en el jardin.',
    'El perro duerme en la alfombra.',
    'La gata juega en el sofa.'
]

# Creamos un vocabulario de palabras a tokens.
tokenizer = tf.keras.preprocessing.text.Tokenizer(filters='', split=' ')
tokenizer.fit_on_texts(corpus)

# Convertimos las oraciones a secuencias de enteros.
# `texts_to_sequences` convierte cada palabra en su token (número).
sequences = tokenizer.texts_to_sequences(corpus)
word_index = tokenizer.word_index
vocab_size = len(word_index) + 1  # +1 para el token de padding

# Determinamos la longitud máxima de las secuencias para el padding.
max_sequence_length = max(len(s) for s in sequences)

# `pad_sequences` asegura que todas las secuencias tengan la misma longitud.
padded_sequences = tf.keras.preprocessing.sequence.pad_sequences(sequences, maxlen=max_sequence_length, padding='post')

# Separamos los datos de entrada (X) y la salida esperada (y).
# El modelo aprenderá a predecir la siguiente palabra en la secuencia.
X = padded_sequences[:, :-1]
y = padded_sequences[:, 1:]

print("Datos de entrada (X):")
print(X)
print("\nDatos de salida (y):")
print(y)

# --- 2. Definir los componentes clave de la arquitectura Transformer ---
# Esta es la parte del "neuro-algoritmo".

# a) Positional Encoding (Codificación Posicional)
# Los Transformers no tienen un orden inherente. Esta capa inyecta
# información sobre la posición de cada palabra en la secuencia.
class PositionalEncoding(layers.Layer):
    def __init__(self):
        super(PositionalEncoding, self).__init__()
    
    def get_angles(self, pos, i, d_model):
        angle_rates = 1 / np.power(10000, (2 * (i // 2)) / np.float32(d_model))
        return pos * angle_rates

    def call(self, inputs):
        d_model = inputs.shape[-1]
        max_pos_length = inputs.shape[-2]
        pos = tf.range(max_pos_length, dtype=tf.float32)[:, tf.newaxis]
        i = tf.range(d_model, dtype=tf.float32)[tf.newaxis, :]
        
        angles = self.get_angles(pos, i, d_model)
        
        sines = tf.math.sin(angles[:, 0::2])
        cosines = tf.math.cos(angles[:, 1::2])
        
        pos_encoding = tf.concat([sines, cosines], axis=-1)
        pos_encoding = pos_encoding[tf.newaxis, ...]
        
        return inputs + pos_encoding

# b) Multi-Head Attention (Mecanismo de Auto-atención)
# Esta es la clave del Transformer. Permite que el modelo
# evalúe la importancia de cada palabra en relación con las demás.
def multi_head_attention_block(d_model, num_heads, dff, dropout_rate):
    attn_layer = layers.MultiHeadAttention(num_heads=num_heads, key_dim=d_model, dropout=dropout_rate)
    dense_relu_layer = tf.keras.Sequential([
        layers.Dense(dff, activation='relu'),
        layers.Dense(d_model)
    ])
    layernorm1 = layers.LayerNormalization(epsilon=1e-6)
    layernorm2 = layers.LayerNormalization(epsilon=1e-6)
    
    def apply_block(x):
        attn_output = attn_layer(query=x, value=x, key=x)
        x = layernorm1(x + attn_output)
        ffn_output = dense_relu_layer(x)
        x = layernorm2(x + ffn_output)
        return x
    
    return apply_block

# --- 3. Construir el modelo completo de Transformer ---
def create_transformer_model(vocab_size, max_seq_len, d_model, num_heads, dff, num_layers, dropout_rate):
    inputs = layers.Input(shape=(max_seq_len-1,), dtype=tf.int32)
    
    # 1. Capa de Embedding
    embedding_layer = Embedding(input_dim=vocab_size, output_dim=d_model)
    x = embedding_layer(inputs)
    
    # 2. Capa de Positional Encoding
    x = PositionalEncoding()(x)
    
    # 3. Múltiples bloques de Transformer
    for _ in range(num_layers):
        x = multi_head_attention_block(d_model, num_heads, dff, dropout_rate)(x)
    
    # 4. Capa de salida para la predicción de la siguiente palabra
    output = layers.Dense(vocab_size)(x)
    
    return Model(inputs=inputs, outputs=output)

# Hiperparámetros del modelo
d_model = 64        # Dimensión de la representación interna
num_heads = 4       # Número de "cabezas" de atención
dff = 128           # Dimensión de la red neuronal de feed-forward
num_layers = 2      # Número de bloques de atención
dropout_rate = 0.1

# Crear el modelo
model = create_transformer_model(
    vocab_size=vocab_size,
    max_seq_len=max_sequence_length,
    d_model=d_model,
    num_heads=num_heads,
    dff=dff,
    num_layers=num_layers,
    dropout_rate=dropout_rate
)

# Compilar el modelo
# Usamos 'sparse_categorical_crossentropy' porque predecimos el siguiente token.
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

# Imprimir el resumen del modelo para ver las capas.
print("\n--- Resumen del Modelo Transformer ---")
model.summary()

# --- 4. Entrenar el modelo ---
# Entrenamos el modelo para que aprenda las relaciones entre palabras.
print("\n--- Entrenando el modelo ---")
model.fit(X, y, epochs=50)

# --- 5. Demostración de generación de texto ---
# El modelo ahora puede generar texto.

def generate_text(model, start_text, num_words_to_generate, tokenizer, max_seq_len):
    # Convertir el texto de inicio a tokens.
    input_sequence = tokenizer.texts_to_sequences([start_text])[0]
    
    for _ in range(num_words_to_generate):
        # Aseguramos que la secuencia tenga la longitud correcta para el modelo.
        input_padded = tf.keras.preprocessing.sequence.pad_sequences([input_sequence], maxlen=max_seq_len-1, padding='post')
        
        # Realizamos la predicción. El modelo predice la distribución de probabilidad
        # para la siguiente palabra.
        predictions = model.predict(input_padded)[0][-1]
        predicted_token_index = np.argmax(predictions)
        
        # Obtenemos la palabra del token.
        predicted_word = tokenizer.index_word.get(predicted_token_index, "")
        
        # Si la palabra predicha es un token de padding, paramos.
        if predicted_word == "":
            break
        
        # Agregamos la palabra predicha a la secuencia de entrada para la siguiente predicción.
        input_sequence.append(predicted_token_index)
        
    generated_text = ' '.join([tokenizer.index_word.get(i, "") for i in input_sequence])
    return generated_text

# Generar un texto de ejemplo
start_sentence = "El gato"
generated_sentence = generate_text(model, start_sentence, 5, tokenizer, max_sequence_length)
print(f"\nFrase inicial: '{start_sentence}'")
print(f"Frase generada por el modelo: '{generated_sentence}'")
